[Skia] Check if glFence is available before using it

carlosgcampos opened the pull request:
#### c4b3e26f6881ae3d97a0ac8f1cb09a105cd95522
<pre>
[Skia] Check if glFence is available before using it
<a href="https://bugs.webkit.org/show_bug.cgi?id=271376">https://bugs.webkit.org/show_bug.cgi?id=271376</a>

Reviewed by Miguel Gomez.

If not available we need to call GrDirectContext::submit() with
GrSyncCpu=kYes. This patch adds a helper class GLFence to wrap it, that
can be extended in the future to support nvidia fences if needed too.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PlatformPlayStation.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::glExtensions const):
* Source/WebCore/platform/graphics/egl/GLContext.h:
* Source/WebCore/platform/graphics/egl/GLFence.cpp: Added.
(WebCore::GLFence::create):
(WebCore::GLFence::GLFence):
(WebCore::GLFence::~GLFence):
(WebCore::GLFence::wait):
* Source/WebCore/platform/graphics/egl/GLFence.h: Added.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::AcceleratedBuffer::beginPainting):
(Nicosia::AcceleratedBuffer::completePainting):
(Nicosia::AcceleratedBuffer::waitUntilPaintingComplete):
(Nicosia::AcceleratedBuffer::~AcceleratedBuffer): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:

Canonical link: <a href="https://commits.webkit.org/276463@main">https://commits.webkit.org/276463@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a10bf0143115d1e849120752ca281314be81c9e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47210 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47391 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40742 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27872 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21221 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45314 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20903 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17838 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18329 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39671 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2784 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/40976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39961 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49051 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19700 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/16266 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43753 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21022 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42499 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21359 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6181 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20695 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

magomez commented on Source/WebCore/platform/graphics/egl/GLFence.cpp:
```diff
@@ -0,0 +1,65 @@
+        return nullptr;
+
+    const auto& extensions = context->glExtensions();
+    if (extensions.APPLE_sync) {
```
Fences should be available on GLES3 contexts, or on GLES2 contexts that have GL_APPLE_sync extensions.
This should be something like
if (context->version() >= 300 || extensions.APPLE_sync)

carlosgcampos committed:
[Skia] Check if glFence is available before using it
https://bugs.webkit.org/show_bug.cgi?id=271376

Reviewed by Miguel Gomez.

If not available we need to call GrDirectContext::submit() with
GrSyncCpu=kYes. This patch adds a helper class GLFence to wrap it, that
can be extended in the future to support nvidia fences if needed too.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PlatformPlayStation.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::glExtensions const):
* Source/WebCore/platform/graphics/egl/GLContext.h:
* Source/WebCore/platform/graphics/egl/GLFence.cpp: Added.
(WebCore::GLFence::create):
(WebCore::GLFence::GLFence):
(WebCore::GLFence::~GLFence):
(WebCore::GLFence::wait):
* Source/WebCore/platform/graphics/egl/GLFence.h: Added.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::AcceleratedBuffer::beginPainting):
(Nicosia::AcceleratedBuffer::completePainting):
(Nicosia::AcceleratedBuffer::waitUntilPaintingComplete):
(Nicosia::AcceleratedBuffer::~AcceleratedBuffer): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:

Canonical link: https://commits.webkit.org/276463@main
